jwt-bcp/draft-sheffer-oauth-jwt-bcp.md

```diff
@@ -97,89 +134,186 @@ updates that apply to this document.
+the application, it SHOULD consider the JWS to be invalid."
+
+Therefore, applications MUST only allow the use of cryptographically current algorithms
+that meet the security requirements of the application.
+This set will vary over time as new algorithms are introduced
+and existing algorithms are deprecated due to discovered cryptographic weaknesses.
+Applications must therefore be designed to enable cryptographic agility.
+
+That said, if a JWT is cryptographically protected by a transport layer, such as TLS,
+using cryptographically current algorithms, there may be no need to apply another layer of
+cryptographic protections to the JWT.
+In such cases, the use of the "none" algorithm can be perfectly acceptable.
+
+## Validate All Cryptographic Operations ## {#validate-crypto}
+
+All cryptographic operations used in the JWT MUST be validated and the entire JWT MUST be rejected
+if any of them fail to validate.
+This is true not only of JWTs with a single set of Header Parameters
+but also for Nested JWTs, in which both outer and inner operations MUST be validated
+using the keys and algorithms supplied by the application.
 
 <!-- See draft-ietf-oauth-jwsreq-13, sec. 6.1 and 6.2. -->
 
-## Distinguishing between Different Applications of JWTs
+## Validate Cryptographic Inputs ## {#validate-inputs}
+
+Some cryptographic operations, such as Elliptic Curve Diffie-Hellman key agreement
+("ECDH-ES") take inputs that may contain invalid values, such as points not on the specified elliptic curve
+or other invalid points.
+Either the JWS/JWE library itself must validate these inputs before using them
+or it must use underlying cryptographic libraries that do so (or both!).
+
+## Ensure Cryptographic Keys have Sufficient Entropy {#key-entropy}
+
+The Key Entropy and Random Values advice in Section 10.1 of [RFC7515] and
+the Password Considerations in Section 8.8 of [RFC7518]
+MUST be followed.
+
+## Use UTF-8 ## {#use-utf8}
 
+[RFC7515], [RFC7516], and [RFC7519] all specify that UTF-8 be used for encoding and decoding JSON
+used in Header Parameters and JWT Claims Sets.
+Implementations and applications MUST do this, and not use other Unicode encodings for these purposes.
+
+## Validate Issuer and Subject ## {#validate-iss-sub}
+
+When a JWT contains an "iss" (issuer) claim, the application MUST validate that the cryptographic keys
+used for the cryptographic operations in the JWT belong to the issuer.
+If they do not, the application MUST reject the JWT.
+
+Similarly, when the JWT contains a "sub" (subject) claim, the application MUST validate that
+the subject value corresponds to a valid subject and/or issuer/subject pair at the application.
+This may include confirming that the issuer is trusted by the application.
+If the issuer, subject, or the pair are invalid, the application MUST reject the JWT.
+
+## Use and Validate Audience ## {#use-aud}
+
+If the same issuer can issue JWTs that can be used by more than one relying party or application,
```

Comment:
The wording "can be used" is quite vague. Maybe we should acknowledge this fact, and add that in case of doubt, the issuer SHOULD include an "aud" claim anyway.